sw/lib/sw/device/silicon_creator/rom_patch.c

```diff
@@ -0,0 +1,280 @@
+    }
+
+    next_patch_offset += current_patch_size * 4;
+  } while (current_patch_size > 0 &&
+           next_patch_offset < OTP_CTRL_PARAM_NUM_SW_CFG_WINDOW_WORDS * 4);
+
+  if (latest_patch_offset == SIZE_MAX) {
+    return kRomPatchInvalidAddr;
+  }
+
+  patch_header = otp_read32(kRomPatchBaseAddr + latest_patch_offset);
+  OT_DISCARD(rom_printf("Latest patch header 0x%x\n", patch_header));
+
+  return (uintptr_t)(kRomPatchBaseAddr + latest_patch_offset);
+}
+
+static OT_WARN_UNUSED_RESULT rom_error_t rom_patch_verify_sig(
+    const rom_patch_t *patch, const hmac_digest_t *patch_digest) {
+  /*
+   * TODO: Verify that the loaded code matches the OTP signature.
+   */
+  return kErrorOk;
+}
+
+static OT_WARN_UNUSED_RESULT rom_error_t
+rom_patch_remap(const rom_patch_t *patch) {
+  for (uint32_t i = 0; i < RV_CORE_IBEX_PARAM_NUM_REGIONS; i++) {
+    const rom_patch_match_regs_t *match = &patch->table[i];
+    // If patch is not enabled, skip it.
+    if (!rom_patch_region_enabled(match)) {
+      continue;
+    }
+
+    uint32_t patch_size_bytes = (uint32_t)(rom_patch_region_size(match) << 2);
+    uint32_t m_base = rom_patch_region_m_base(match);
+    uint32_t r_base = rom_patch_region_r_base(match);
+    uint32_t mask =
+        (m_base & ~(patch_size_bytes - 1)) | ((patch_size_bytes - 1) >> 1);
+
+    sec_mmio_write32(kIbexBase + IBEX_IBUS_ADDR_MATCHING_REG(i), mask);
+    sec_mmio_write32(kIbexBase + IBEX_IBUS_REMAP_ADDR_REG(i), r_base);
+    sec_mmio_write32(kIbexBase + IBEX_IBUS_ADDR_EN_REG(i), 1);
+    if (rom_patch_region_locked(match)) {
+      sec_mmio_write32(kIbexBase + IBEX_IBUS_REGWEN_REG(i), 0);
+    }
+
+    OT_DISCARD(rom_printf("Configured Ibex remapping reg 0x%x->0x%x %x\n", mask,
+                          r_base, rom_patch_region_locked(match)));
+  }
+
+  icache_invalidate();
+
+  return kErrorOk;
+}
+
+OT_WARN_UNUSED_RESULT rom_error_t
+rom_patch_apply(const uintptr_t patch_addr, hmac_digest_t *const patch_digest) {
+  size_t patch_code_offset = kRomPatchRegionPreambleSize;
+  uint32_t patch_preamble_bytes[kRomPatchRegionPreambleSize];
+  rom_patch_t *patch;
+
+  hmac_sha256_init();
+
+  // Read the patch preamble from OTP.
+  otp_read(patch_addr, patch_preamble_bytes, kRomPatchRegionPreambleSize);
+  patch = (rom_patch_t *)(patch_preamble_bytes);
+
+  // The first header byte (LOCK_VALID & PROGRAM_START) is not signed.
+  hmac_sha256_update(patch_preamble_bytes + 1, kRomPatchRegionPreambleSize - 1);
+
+  // The patch code size must not be 0.
+  size_t patch_code_size_bytes = rom_patch_code_size(patch);
+  HARDENED_CHECK_NE(patch_code_size_bytes, 0);
+
+  // The remapping base address is the first entry in the match table.
+  const rom_patch_match_regs_t *match = &patch->table[0];
+  uint32_t remap_addr = rom_patch_region_r_base(match);
+
+  /*
+   * Read the whole patch section, dword by dword, and copy it to the
+   * remapped address.
+   */
+  for (size_t i = 0; i < patch_code_size_bytes; i += 4) {
```

Comment:
Here you are assuming that the size is in bytes and a multiple of 4. This is true because of the way the doc works but not obvious when reading the code. I would suggest that `rom_patch_code_size` be renamed `rom_patch_code_size_words`, returns a size in words, and then multiply by 4 in this function.

Reply:
Done.